### DISCUSSION TITLE
Connexion 3 alpha release is out now! 🎉

### DISCUSSION THREAD

**Discussion by RobbeSneyders** [spec-first/connexion#1660] [Announcements]
We've been working on Connexion 3 for about a year now, and we're happy to finally be able to release a first alpha release. The goal of the alpha release is to provide users the ability to test the new version and provide feedback while we further polish the codebase and update the documentation.

#### 🛠️ Connexion 3 brings some major changes:

- Connexion can now be used as middleware to supercharge any ASGI or WSGI compatible framework.
- Aiohttp support has been dropped in favor of an ASGI compatible `AsyncApp` based on Starlette which is almost a drop-in replacement for the FlaskApp.
- Connexion functionality is now pluggable by adding or removing middleware.
- Validation is now pluggable by content type, solving longstanding issues regarding endpoints with multiple content types and providing the ability to plug in your own validation logic.

But these are only the highlights. There's a whole list of changes, fixes, and improvements, which you can read about in the changelog.

#### 📃 Documentation

We still need to update the documentation for Connexion 3, but in the meantime, we've created a single page which provides high level documentation to get started with the new version [here](https://connexion.readthedocs.io/en/latest/v3.html).

#### 📣 Feedback

We would really love to hear from you. By trying Connexion 3 now and providing feedback, you get the opportunity to make sure that updating your application will be as painless as possible, while helping to create a more stable GA release.
- For questions, comments, and feedback, please use this thread
- For issues specific to the alpha version, please create a new issue and mark it with the Connexion 3 milestone
- For missing documentation, please comment on https://github.com/spec-first/connexion/issues/1531

**Comment by perrystallings** [comment]
Started testing the starlette app and found 2 issues:

- release is missing the asgiref dependency
- running AsyncApp still requires a flask install (is this intentional?)

**Reply by RobbeSneyders** [reply]
Thanks for the feedback @perrystallings!

> - release is missing the asgiref dependency

Submitted #1667 to fix this

> - running AsyncApp still requires a flask install (is this intentional?)

This is indeed not intentional, you should be able to run the AsyncApp without installing Flask. Submitted #1668 to fix this.

We'll release this as a new version asap.

**Reply by perrystallings** [reply]
Thanks!

**Reply by RobbeSneyders** [reply]
These changes have been released as 3.0.0a2.

**Comment by perrystallings** [comment]
Further exploring the AsyncApp, why does AsyncApp.middleware_app.app return a Starlette Router rather than a Starlette app? (FlaskApp.middleware_app.app returns a flask app and the old AioHttp did the same)

Context: trying to configure opentelemetry without writing custom a connexion middleware. Unfortunately this seems to require a starlette app rather than a Router

- Github: https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-starlette
- Docs: https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/starlette/starlette.html

**Reply by RobbeSneyders** [reply]
Thanks for this! This is exactly the feedback we're looking for: real-life use cases that we need to cater to.

---

The `AsyncApp` is built on top of a Starlette `Router` because a Starlette `App` is a `Router` wrapped with a middleware stack. Which is what Connexion now is as well. It's a lot simpler to not have multiple middleware stacks wraped by each other. The default middleware stack of Starlette would also duplicate some of the behavior of Connexion (such as error handling), which we don't want to happen there.

---

I had a quick look at your links, and the good news is that the opentelemetry instrumentation also provides a [generic ASGI middleware](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/asgi/asgi.html), which is exactly what Connexion 3 now enables. You should be able to plug it into the middleware stack as follows (untested):

```
from connexion import AsyncApp, ConnexionMiddleware
from opentelemetry.instrumentation.asgi import OpenTelemetryMiddleware

middleware_stack = ConnexionMiddleware.default_middlewares + [OpenTelemetryMiddleware]

app = AsyncApp(__name__, middlewares=middleware_stack)
app.add_api(...)
```

It's not completely clear yet to me what the advantage of the Starlette specific instrumentator is.

---

Note that if you want to pass arguments to middleware you add to the stack, you can use a partial.

```
from functools import partial

middleware = partial(OpenTelemetryMiddleware, *args, **kwargs)
middleware_stack = ConnexionMiddleware.default_middlewares + [OpenTelemetryMiddleware]

app = AsyncApp(__name__, middlewares=middleware_stack)
```

We currently use the [same interface to specify middleware as Django](https://docs.djangoproject.com/en/4.1/topics/http/middleware/#activating-middleware), which allows a lot of flexibility. This is needed because the order of the middleware matters.

We might want to add an `add_middleware()` method similar to [Starlette's](https://github.com/encode/starlette/blob/b3add7e7ac61c16af953644e2778ea32d9e125ba/starlette/applications.py#L132) to enable simple use cases though. This is something I would still consider for the 3.0 release.

We probably should at least expose the `default_middlewares` via the app directly instead of via the `ConnexionMiddleware`.

**Reply by perrystallings** [reply]
Thanks for the context! Just tested the configuration and it works!

**Comment by SpudInNZ** [comment]
First off, congrats to the team for getting to the 3.0.0 release.

Is there an expected migration path for 2.x projects built upon AioHttp? I'm guessing not, since AioHttp (to the best of my knowledge) is not ASGI compatioble and the team there seems to have given up on development of the 4.x branch. 

Or do I have this wrong?

**Reply by RobbeSneyders** [reply]
Thanks @SpudInNZ!

As far as I'm aware, there are indeed no plans for AioHttp to support the ASGI spec, which unfortunately means connexion can no longer support it.

I see 2 possible migration paths:
- Switch to Connexion's `AsyncApp`. I think this should be possible if you're not heavily dependent on the AioHttp ecosystem. We're happy to help make this as painless as possible. So if you can provide feedback on the migration, we can see if there's ways for us to make this easier or at least provide documentation.
- Switch to a Connexion 2 fork such as https://github.com/athenianco/especifico by @vmarkovtsev

**Reply by SpudInNZ** [reply]
@RobbeSneyders 

Thanks for the reply. In truth it's difficult for us to make the case for migration. However we may start new projects with the new stack.

**Comment by drummerwolli** [comment]
Hi @RobbeSneyders , i created an issue, but i cant tag it with the 3.0.0 milestone: https://github.com/spec-first/connexion/issues/1670 ...

congratz on the release progress! looking forward!

**Comment by SpudInNZ** [comment]
According to the docs, "Connexion now exposes context, operation, receive, scope as global request-level context". 
How is this exposed to an endpoint handler? I can't find anything in the example projects. For example, in the implementation of:

```
@StarletteDecorator()
async def get_greeting(name: str) -> str:
    return f"Hello {name}!"

```
.. how would I access the request headers?

Connexion 2.x `app.add_api()` supported the `pass_context_arg_name` parameter, this was extremely useful. Could this be supported in 3.x as well, passing in (presumably) a Starlette Request to api handlers? Or are we expected to use custom middleware to achieve this now?

EDIT: After some digging around I've found I can use

```
import typing as t
from connexion.decorators import StarletteDecorator
from connexion.context import scope, receive
import starlette.requests


@StarletteDecorator()
async def get_greeting(name: str) -> str:
    request = starlette.requests.Request(scope, receive)
    hdrs = request.headers  # for example

    return f"Hello {name}!"

```
Is this what I am supposed to do?

**Reply by RobbeSneyders** [reply]
We indeed still need to document this behavior. Working on the documentation as we speak :slightly_smiling_face: 

You can access all these objects in the request-level context by importing from `connexion.context`:
```
from connexion.context import context, operation, request, receive, scope
```

The `request` is also still available as 
```
from connexion import request
```
but that is now a Starlette `Request` instead of a Flask one.

The `pass_context_arg_name` parameter has been removed, but the mechanism hasn't. So ou can no longer define the name, but you can still get the connexion `context` in your handler function via the hardcoded `context_` parameter name.
```
def handler(context_)
    pass
```
I'm wondering if we still need this now that the `context` is available globally as well though.

**Comment by jagadeeshgarvis** [comment]
Thanks for this work! I am migrating one of our applications to connexion 3 (with starlette) from connexion 2 (with flask).

I am facing some difficulties is accessing form or files data. I am able to access query_params without any issues using `connexion.requests`

I am able to port almost everything except for this error in reading uploaded files as multipart form field. 

`await connexion.request.form() `or `await connexion.request.files() `is making the app go into a limbo. 

Even `form = await connexion.request._starlette_request.form()` this doesn't seems to be working

Even if I place this inside a try, catch, there are no error messages and the app isn't returning anything.

**Reply by RobbeSneyders** [reply]
Hi @jagadeeshgarvis, thanks for the feedback!

I can indeed reproduce this. What is working, is leveraging the [automatic parameter handling](https://connexion.readthedocs.io/en/stable/request.html#automatic-parameter-handling), which lets you receive the data as arguments to your view function.

The issue is that because of this automatic parameter handling, we already consume the stream before you access the request. This is a bug though and I'll work on a fix soon.

**Reply by jagadeeshgarvis** [reply]
Thanks @RobbeSneyders for the quick response. Really looking forward to this :)

**Reply by jagadeeshgarvis** [reply]
Also @RobbeSneyders Is there a way, we can access the files (before this bug is fixed). For example, before access any parameters, if we try calling the files/forms as the first thing, will it work? If you can share some insights regarding the same, it will be really helpful. 

We may need something to get our current app working with connexion 3.0

Thanks!

**Reply by RobbeSneyders** [reply]
You can leverage the automatic parameter handling.

If this is your spec:
```
operationId: main.post
requestBody:
  content:
    multipart/form-data:
      schema:
        type: object
        properties:
          document:
            type: string
            format: binary
```

You can get the files by adding a `document` argument to your `main.post` view function:
```
def post(document):
```
And they will automatically be injected.

**Reply by jagadeeshgarvis** [reply]
@RobbeSneyders Not all fields are available using this method as well..

Let's say, I have my spec like this:
```
          multipart/form-data:
            schema:
              type: object
              properties:
                file:
                  type: string
                  format: binary
                  description: Event file to upload
                filetype:
                  type: string
                  $ref: "#/components/schemas/EventUploadType"
              required:
                - file
                - filetype

```

I am able to access: `file`  but I am not able to access `filetype`.

Also just for context, when I am debugging, this field `filetype` was available in `_form` field of the ASGI Request till we reach this point in. code:

```
class AsyncParameterDecorator(BaseParameterDecorator):
    def __call__(self, function: t.Callable) -> t.Callable:
        unwrapped_function = unwrap_decorators(function)
        arguments, has_kwargs = inspect_function_arguments(unwrapped_function)

        @functools.wraps(function)
        async def wrapper(request: ASGIRequest) -> t.Any:
            request_body = self._maybe_get_body(
                request, arguments=arguments, has_kwargs=has_kwargs
            )

            while asyncio.iscoroutine(request_body):
                request_body = await request_body

            kwargs = prep_kwargs(
                request,
                request_body=request_body,
                files=await request.files(),
                arguments=arguments,
                has_kwargs=has_kwargs,
                sanitize=self.sanitize_fn,
            )

            return await function(**kwargs)

        return wrapper

```
Pls See the value in debugger:
<img width="1598" alt="Screenshot 2023-04-05 at 7 58 39 PM" src="https://user-images.githubusercontent.com/121790234/230112253-c20111f0-98ad-4c83-98d1-314e1f8db906.png">

But this field is not accessible in the actual `post` view function. I had tried passing them as arguments and also tried accessing the `_form`. 

Anyway to overcome this?

I am also wondering if this could be related to the name of the parameter: `filetype` is this causing any issues?

**Reply by RobbeSneyders** [reply]
You're right. With Swagger 2, this works since form data is defined as parameters in the specification. With OpenAPI 3, form data is defined in the body, so we pass in the whole body.

This will work:
```
def post(file, body):
    # body = {"filetype": "..."}
```

But I agree that this is not very consistent. I think we should either
- Copy the Swagger behavior for OpenAPI and pass the top level fields in separately.
  (This might impact the behavior of `strict_validation` as well.)
- Don't pass in the files separately for OpenAPI either, and just include them in the body

I'll have to look how this was done on Connexion 2.X before deciding on how to move forward.
CC: @Ruwann

**Reply by jagadeeshgarvis** [reply]
Thanks for quick response... and helping here.... now I am able to access it from `body`. Posting it here so that it may help others as well.

<img width="1071" alt="Screenshot 2023-04-05 at 8 37 01 PM" src="https://user-images.githubusercontent.com/121790234/230123397-36a97cf4-c98e-4a3d-884e-1bd65adc4963.png">

**Reply by RobbeSneyders** [reply]
Yes that should work. Both options mentioned above will change this interface and break your code though. But I'll keep you updated here so you can change it when the fix is released.

**Reply by RobbeSneyders** [reply]
@Ruwann I checked and this is the same behavior as we had on Connexion 2 (see also https://github.com/spec-first/connexion/discussions/1679)

I would propose to go with a single body argument and include the files in there under there respective key. This is consistent with other content-types and we can still make this more user friendly in the future by providing parsing functionality (https://github.com/spec-first/connexion/issues/1190).

**Comment by rashmi-mi** [comment]
Ran into a peculiar issue - we have 2 APIs defined in OpenAPI yml - 
/exceptions/{exception-id}
/exceptions/dates

Since /exceptions/{exception-id} is defined before the /exceptions/dates API, Starlette routing picks up the first API all the time. This used to work with connexion 2.0. I am guessing the right fix is changing the order but wanted to bring it to your notice here.

**Reply by RobbeSneyders** [reply]
Yes, we are aware. This is [documented in the change log](https://connexion.readthedocs.io/en/latest/v3.html#:~:text=Route%20priority%20changed.%20The%20most%20specific%20route%20should%20now%20be%20defined%20first%20in%20the%20specification.) as a breaking change. We'll make sure to highlight this in the routing documentation again.

**Reply by rashmi-mi** [reply]
Thanks @RobbeSneyders! Missed that one.

**Comment by rashmi-mi** [comment]
Hi @RobbeSneyders  Do you think CORS support be made part of the default implementation since the order in which the middleware is inserted matters? 
If CORSMiddleware is the last one added, the swagger implementation returns with a 405 for OPTIONS method. Making it the first Middleware before the default list works fine. Not sure which other middleware would interfere with the CORS Middleware.

**Reply by RobbeSneyders** [reply]
We will not make it default, since it is not always desired. 

Adding middleware at specific locations will become easer with #1683 which I just opened yesterday. It adds an `add_middleware` interface to add middleware to the stack, which supports different logical locations. 

The CORS middleware needs to be inserted before the `RoutingMiddleware`, so you'll be able to do:
```
from connexion.middleware import MiddlewarePosition

app.add_middleware(CorsMiddleware, position=MiddlewarePosition.BEFORE_ROUTING)
```

**Reply by rashmi-mi** [reply]
This is great, thanks @RobbeSneyders!

**Reply by RobbeSneyders** [reply]
This has been released as [3.0.0a5](https://github.com/spec-first/connexion/releases/tag/3.0.0a5)

**Reply by rashmi-mi** [reply]
Thanks!

**Reply by RobbeSneyders** [reply]
Latest release is 3.0.0a6 already, which contains some additional fixes related to this functionality.

**Comment by ergo** [comment]
Hi, thank you for all the hard work on Connexion. Is there a ballpark date for beta/rc?

**Comment by SpudInNZ** [comment]
It seems to have gone awfully quiet with this project, can we get an update please?

**Reply by RobbeSneyders** [reply]
Hi @SpudInNZ,
Both @Ruwann and I had less time to spend on connexion the previous weeks, but we are still moving forward (albeit slowly), which also allows for additional feedback on the Alpha release.

We are ironing out some final issues which we want to resolve (eg. https://github.com/spec-first/connexion/issues/1709) and adding documentation (#1701, #1706), which unfortunately takes a bit longer than hoped.

If people want to help, they can triage issues or help write documentation (https://github.com/spec-first/connexion/issues/1531)

**Comment by danderwald** [comment]
Hi,

first things first: congrats!

I was wondering if it would be possible to easily migrate to v3 and migrate flask to quart (=keep the migration fairly short but enables async coding).

According to the docs, this will not be possible, because automatic routing / parameter injection / response serialization will not be available. Can this somehow be added "manually" (without implementing it explicitly :))? or is this not foreseen?

**Reply by RobbeSneyders** [reply]
Hi @danderwald. Is there a reason for you to want to migrate to quart instead of the AsyncApp based on Starlets?

**Reply by danderwald** [reply]
Hi,

pragmatic reasons :) 
- one reason is that quarts api is the same as the flask one (for example for templating / globals /events / ...), so I just would need to replace the imports
- because the API is the  same, my team would not need to learn a new framework
- we would be able to use concurrency (in combination with an asgi based framework/server) without much migration effort ( => to get rid of some blocking IO by exchanging requests with httpx)

**Comment by hedleyroos** [comment]
I don't have a question but I need to congratulate you somehow! Ported a Connexion 2 application that uses Django's ORM behind the scenes in a few hours. So far everything just works.

**Reply by RobbeSneyders** [reply]
Thanks @hedleyroos! Great to hear!

Let us know if you run into any issues.

**Comment by Qreativeai** [comment]
excited for final release

**Comment by RobbeSneyders** [comment]
Connexion 3 is out now! 🎉
Any further discussion can happen on the [new announcement](https://github.com/spec-first/connexion/discussions/1775).
